### PR TITLE
[#79] Implement object commands

### PR DESCRIPTION
**Pull request by fyrchik**
Implement everything except `Search`.

### CODE REVIEW

File: cmd/neofs-cli/modules/object.go [hunk 1]
```diff
@@ -29,3 +130,368 @@ func init() {
...
+			WithPayloadReader(f),
+		client.WithSession(tok))
+	if err != nil {
+		return fmt.Errorf("can't put object: %w", err)
+	}
+
+	cmd.Printf("[%s] Object successfully stored\n", filename)
+	cmd.Printf("  ID: %s\n  CID: %s\n",
+		base58.Encode(oid.ToV2().GetValue()),
+		base58.Encode(cid.ToV2().GetValue()))
```
Comment (alexvanin) [inline cmd/neofs-cli/modules/object.go:188]: With https://github.com/nspcc-dev/neofs-node/pull/98 you can just  `cmd.Printf("  ID: %s\n  CID: %s\n", oid, cid)`

File: cmd/neofs-cli/modules/object.go [hunk 2]
```diff
@@ -29,3 +130,368 @@ func init() {
...
+		return err
+	}
+
+	ctx := context.Background()
+	cli, tok, err := initSession(ctx)
+	if err != nil {
+		return err
+	}
+	obj, err := cli.GetObjectHeader(ctx,
+		new(client.ObjectHeaderParams).WithAddress(objAddr),
```
Comment (alexvanin) [inline cmd/neofs-cli/modules/object.go:274]: Consider parsing `--main-only` flag. If set, then add `.WithMainFields()` option.

File: cmd/neofs-cli/modules/object.go [hunk 3]
```diff
@@ -1,24 +1,125 @@
...
+	objectHeadCmd.Flags().Bool("json", false, "Marshal output in JSON")
+	objectHeadCmd.Flags().Bool("proto", false, "Marshal output in Protobuf")
+
+	objectCmd.AddCommand(objectHashCmd)
+	objectHashCmd.Flags().String("cid", "", "Container ID")
+	_ = objectHashCmd.MarkFlagRequired("cid")
+	objectHashCmd.Flags().String("oid", "", "Object ID")
+	_ = objectHashCmd.MarkFlagRequired("oid")
+	objectHashCmd.Flags().StringSlice("range", []string{}, "Range to take hash from.")
+	objectHashCmd.Flags().String("type", hashSha256, "Hash type. Either 'sha256' (default) or 'tz'")
```
Comment (alexvanin) [inline cmd/neofs-cli/modules/object.go:120]: Default value specified automatically by Cobra, so you don't have to comment about this. 
Now help line looks
```
      --type string     Hash type. Either 'sha256' (default) or 'tz' (default "sha256")
```

File: cmd/neofs-cli/modules/object.go [hunk 4]
```diff
@@ -1,24 +1,125 @@
...
+	_ = objectHeadCmd.MarkFlagRequired("oid")
+	objectHeadCmd.Flags().Bool("json", false, "Marshal output in JSON")
+	objectHeadCmd.Flags().Bool("proto", false, "Marshal output in Protobuf")
+
+	objectCmd.AddCommand(objectHashCmd)
+	objectHashCmd.Flags().String("cid", "", "Container ID")
+	_ = objectHashCmd.MarkFlagRequired("cid")
+	objectHashCmd.Flags().String("oid", "", "Object ID")
+	_ = objectHashCmd.MarkFlagRequired("oid")
+	objectHashCmd.Flags().StringSlice("range", []string{}, "Range to take hash from.")
```
Comment (alexvanin) [inline cmd/neofs-cli/modules/object.go:119]: Add an example `in a form of 1:2,3:4`

File: cmd/neofs-cli/modules/object.go [hunk 5]
```diff
@@ -29,3 +130,368 @@ func init() {
...
+		new(client.DeleteObjectParams).WithAddress(objAddr),
+		client.WithSession(tok))
+	if err != nil {
+		return err
+	}
+
+	cmd.Println("Object removed successfully.")
+	cmd.Printf("  ID: %s\n  CID: %s\n",
+		base58.Encode(objAddr.GetObjectID().ToV2().GetValue()),
+		base58.Encode(objAddr.GetContainerID().ToV2().GetValue()))
```
Comment (alexvanin) [inline cmd/neofs-cli/modules/object.go:213]: See message about available stringer interface.

File: cmd/neofs-cli/modules/object.go [hunk 6]
```diff
@@ -29,3 +130,368 @@ func init() {
...
+		}
+		err := ioutil.WriteFile(filename, bs, os.ModePerm)
+		if err != nil {
+			return err
+		}
+		cmd.Printf("[%s] Header successfully saved.", filename)
+	}
+
+	cmd.Printf("CID: %s\n", base58.Encode(obj.GetContainerID().ToV2().GetValue()))
+	cmd.Printf("OID: %s\n", base58.Encode(obj.GetID().ToV2().GetValue()))
```
Comment (alexvanin) [inline cmd/neofs-cli/modules/object.go:469]: See message about available stringer interface.

File: cmd/neofs-cli/modules/object.go [hunk 7]
```diff
@@ -29,3 +146,438 @@ func init() {
...
+		cmd.Printf("[%s] Header successfully saved.", filename)
+	}
+
+	cmd.Printf("ID: %s\n", obj.GetID())
+	cmd.Printf("CID: %s\n", obj.GetContainerID())
+	cmd.Printf("Owner: %s\n", obj.GetOwnerID())
+	cmd.Printf("CreatedAt: %d\n", obj.GetCreationEpoch())
+	cmd.Printf("Size: %d\n", obj.GetPayloadSize())
+	cmd.Printf("HomoHash: %s\n", hex.EncodeToString(obj.GetPayloadHomomorphicHash().GetSum()))
+	cmd.Printf("Checksum: %s\n", hex.EncodeToString(obj.GetPayloadChecksum().GetSum()))
```
Comment (alexvanin) [inline cmd/neofs-cli/modules/object.go:562]: Let's add object type information there. Something like this:
```
	switch obj.GetType() {
	case object.TypeRegular:
		cmd.Println("Type: regular")
	case object.TypeTombstone:
		cmd.Println("Type: tombstone")
	case object.TypeStorageGroup:
		cmd.Println("Type: storage group")
	default:
		cmd.Println("Type: unknown")
	}
```

**Review by alexvanin — CHANGES_REQUESTED** [review]

File: cmd/neofs-cli/modules/object.go [hunk 8]
```diff
@@ -29,3 +146,438 @@ func init() {
...
+	case noleaf:
+		fs.AddNonLeafFilter()
+	}
+
+	return fs, nil
+}
+
+func parseObjectAttrs(cmd *cobra.Command) ([]*object.Attribute, error) {
+	raw := cmd.Flag("attributes").Value.String()
+	if len(raw) == 0 {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply cmd/neofs-cli/modules/object.go:440]: Fixed.

**Review by alexvanin — APPROVED** [review]

**Pull request merged**